pubsub: GCP PubSub needs a MaxMessages param in ReceiveBatch

Currently it's giving this error from the conformance test:

>     --- FAIL: TestConformance/TestSendReceive (2.15s)
>         drivertest.go:91: rpc error: code = InvalidArgument desc = A required argument is missing in the request: (argument="max_messages").

I'm inclined to add a maxMessages argument to ReceiveBatch and possibly have the concrete API automatically adjust that depending on how quickly messages are being acked.

--- a/internal/pubsub/acks_test.go
+++ b/internal/pubsub/acks_test.go
@@ -28,9 +28,14 @@ type ackingDriverSub struct {
 	sendAcks func(context.Context, []driver.AckID) error
 }
 
-func (s *ackingDriverSub) ReceiveBatch(ctx context.Context) ([]*driver.Message, error) {
-	ms := s.q
-	s.q = nil
+func (s *ackingDriverSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	if len(s.q) <= maxMessages {
+		ms := s.q
+		s.q = nil
+		return ms, nil
+	}
+	ms := s.q[:maxMessages]
+	s.q = s.q[maxMessages:]
 	return ms, nil
 }
 

--- a/internal/pubsub/driver/driver.go
+++ b/internal/pubsub/driver/driver.go
@@ -81,7 +81,7 @@ type Topic interface {
 // Subscription receives published messages.
 type Subscription interface {
 	// ReceiveBatch should return a batch of messages that have queued up
-	// for the subscription on the server.
+	// for the subscription on the server, up to maxMessages.
 	//
 	// If there is a transient failure, this method should not retry but
 	// should return a nil slice and an error. The concrete API will take
@@ -96,7 +96,7 @@ type Subscription interface {
 	// are no messages yet.
 	//
 	// ReceiveBatch should be safe for concurrent access from multiple goroutines.
-	ReceiveBatch(ctx context.Context) ([]*Message, error)
+	ReceiveBatch(ctx context.Context, maxMessages int) ([]*Message, error)
 
 	// SendAcks should acknowledge the messages with the given ackIDs on
 	// the server so that they will not be received again for this

--- a/internal/pubsub/empty_sub_test.go
+++ b/internal/pubsub/empty_sub_test.go
@@ -26,7 +26,7 @@ import (
 // driver mischeviously always returns an empty batch.
 type emptyDriverSub struct{}
 
-func (s *emptyDriverSub) ReceiveBatch(ctx context.Context) ([]*driver.Message, error) {
+func (s *emptyDriverSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
 	return nil, nil
 }
 

--- a/internal/pubsub/gcppubsub/gcppubsub.go
+++ b/internal/pubsub/gcppubsub/gcppubsub.go
@@ -147,10 +147,11 @@ func subscriptionExists(ctx context.Context, client *raw.SubscriberClient, subsc
 }
 
 // ReceiveBatch implements driver.Subscription.ReceiveBatch.
-func (s *subscription) ReceiveBatch(ctx context.Context) ([]*driver.Message, error) {
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
 	req := &pb.PullRequest{
 		Subscription:      s.path,
 		ReturnImmediately: false,
+		MaxMessages:       int32(maxMessages),
 	}
 	resp, err := s.client.Pull(ctx, req)
 	if err != nil {

--- a/internal/pubsub/mempubsub/mem.go
+++ b/internal/pubsub/mempubsub/mem.go
@@ -169,15 +169,12 @@ func (s *subscription) receiveNoWait(now time.Time, max int) []*driver.Message {
 }
 
 const (
-	// Limit on how many messages returned from one call to ReceiveBatch.
-	maxBatchSize = 100
-
 	// How often ReceiveBatch should poll.
 	pollDuration = 250 * time.Millisecond
 )
 
 // ReceiveBatch implements driver.ReceiveBatch.
-func (s *subscription) ReceiveBatch(ctx context.Context) ([]*driver.Message, error) {
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
 	// Check for closed or cancelled before doing any work.
 	if err := s.wait(ctx, 0); err != nil {
 		return nil, err
@@ -186,7 +183,7 @@ func (s *subscription) ReceiveBatch(ctx context.Context) ([]*driver.Message, err
 	// alternative would be complicated by the need to recognize expired messages
 	// promptly.
 	for {
-		if msgs := s.receiveNoWait(time.Now(), maxBatchSize); len(msgs) > 0 {
+		if msgs := s.receiveNoWait(time.Now(), maxMessages); len(msgs) > 0 {
 			return msgs, nil
 		}
 		if err := s.wait(ctx, pollDuration); err != nil {

--- a/internal/pubsub/pubsub.go
+++ b/internal/pubsub/pubsub.go
@@ -169,7 +169,9 @@ func (s *Subscription) getNextBatch(ctx context.Context) error {
 	var msgs []*driver.Message
 	err := retry.Call(ctx, gax.Backoff{}, s.driver.IsRetryable, func() error {
 		var err error
-		msgs, err = s.driver.ReceiveBatch(ctx)
+		// TODO(#691): dynamically adjust maxMessages
+		const maxMessages = 10
+		msgs, err = s.driver.ReceiveBatch(ctx, maxMessages)
 		return err
 	})
 	if err != nil {

--- a/internal/pubsub/pubsub_test.go
+++ b/internal/pubsub/pubsub_test.go
@@ -277,7 +277,7 @@ type failSub struct {
 	calls int
 }
 
-func (t *failSub) ReceiveBatch(ctx context.Context) ([]*driver.Message, error) {
+func (t *failSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
 	t.calls++
 	if t.calls <= nRetryCalls {
 		return nil, errRetry

--- a/internal/pubsub/pubsub_test.go
+++ b/internal/pubsub/pubsub_test.go
@@ -63,11 +63,11 @@ func NewDriverSub() *driverSub {
 	return ds
 }
 
-func (s *driverSub) ReceiveBatch(ctx context.Context) ([]*driver.Message, error) {
+func (s *driverSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
 	for {
 		select {
 		case <-s.sem:
-			ms := s.grabQueue()
+			ms := s.grabQueue(maxMessages)
 			if len(ms) != 0 {
 				return ms, nil
 			}
@@ -78,11 +78,16 @@ func (s *driverSub) ReceiveBatch(ctx context.Context) ([]*driver.Message, error)
 	}
 }
 
-func (s *driverSub) grabQueue() []*driver.Message {
+func (s *driverSub) grabQueue(maxMessages int) []*driver.Message {
 	defer func() { s.sem <- struct{}{} }()
 	if len(s.q) > 0 {
-		ms := s.q
-		s.q = nil
+		if len(s.q) <= maxMessages {
+			ms := s.q
+			s.q = nil
+			return ms
+		}
+		ms := s.q[:maxMessages]
+		s.q = s.q[maxMessages:]
 		return ms
 	}
 	return nil

